### PR TITLE
Dbt autocompletion

### PR DESCRIPTION
**Pull request by fivetran-dmitryboykov**
This PR adds autocompletion feature for all dbt nodes: models, macros, sources.

### CODE REVIEW

File: server/src/completion/DbtCompletionProvider.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    const macroCompletions =
+      jinjaType === JinjaPartType.EXPRESSION_START || jinjaType === JinjaPartType.BLOCK_START
+        ? await this.macroCompletionProvider.provideCompletions(jinjaBeforePositionText)
+        : undefined;
+    if (macroCompletions) {
+      return macroCompletions;
+    }
+
+    return Promise.resolve(undefined);
```
Comment (pgrivachev) [inline server/src/completion/DbtCompletionProvider.ts:44]: `return undefined;`

File: server/src/completion/MacroCompletionProvider.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  static readonly INSTALLED_PACKAGE_SORT_PREFIX = '2';
+
+  constructor(private dbtRepository: DbtRepository) {}
+
+  provideCompletions(jinjaBeforePositionText: string): Promise<CompletionItem[] | undefined> {
+    const macroMatch = MacroCompletionProvider.MACRO_PATTERN.exec(jinjaBeforePositionText);
+    if (macroMatch) {
+      const packageName = macroMatch[0].slice(0, -1);
+      const packageMacros = this.dbtRepository.packageToMacrosMap.get(packageName);
+      return Promise.resolve(packageMacros ? Array.from(packageMacros).map<CompletionItem>(m => this.getMacroCompletionItem(m.name)) : undefined);
```
Comment (pgrivachev) [inline server/src/completion/MacroCompletionProvider.ts:19]: We can make method `async` and get rid of `Promise.resolve(`.

File: server/src/completion/MacroCompletionProvider.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    const wordMatch = MacroCompletionProvider.WORD_PATTERN.exec(jinjaBeforePositionText);
+    if (wordMatch) {
+      return Promise.resolve(
```
Comment (pgrivachev) [inline server/src/completion/MacroCompletionProvider.ts:24]: The same here.

File: server/src/completion/MacroCompletionProvider.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        this.dbtRepository.macros.map<CompletionItem>(m => {
```
Comment (pgrivachev) [inline server/src/completion/MacroCompletionProvider.ts:25]: So we return all macros and they will be filtered by VS Code, right?

File: server/src/completion/MacroCompletionProvider.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+          }
+
+          const label = `(${m.packageName}) ${m.name}`;
+          const insertText = `${m.packageName}.${m.name}`;
+          return this.getMacroCompletionItem(label, insertText, `${MacroCompletionProvider.INSTALLED_PACKAGE_SORT_PREFIX}_${label}`);
+        }),
+      );
+    }
+
+    return Promise.resolve(undefined);
```
Comment (pgrivachev) [inline server/src/completion/MacroCompletionProvider.ts:37]: `return undefined;`

File: server/src/completion/ModelCompletionProvider.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  static readonly CURRENT_PACKAGE_SORT_PREFIX = '1';
+  static readonly INSTALLED_PACKAGE_SORT_PREFIX = '2';
+
+  constructor(private dbtRepository: DbtRepository) {}
+
+  provideCompletions(jinjaBeforePositionText: string): Promise<CompletionItem[] | undefined> {
+    const modelMatch = ModelCompletionProvider.MODEL_PATTERN.exec(jinjaBeforePositionText);
+    if (modelMatch) {
+      const lastChar = jinjaBeforePositionText.charAt(jinjaBeforePositionText.length - 1);
+      return Promise.resolve(
```
Comment (pgrivachev) [inline server/src/completion/ModelCompletionProvider.ts:18]: The same, add `async` to the method signature.

File: server/src/completion/ModelCompletionProvider.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+          return this.getModelCompletionItem(label, insertText, sortOrder);
+        }),
+      );
+    }
+
+    const packageMatch = ModelCompletionProvider.PACKAGE_PATTERN.exec(jinjaBeforePositionText);
+    if (packageMatch) {
+      const [, dbtPackage] = packageMatch;
+      const packageModels = this.dbtRepository.packageToModelsMap.get(dbtPackage);
+      return Promise.resolve(
```
Comment (pgrivachev) [inline server/src/completion/ModelCompletionProvider.ts:35]: The same.

File: server/src/completion/SourceCompletionProvider.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+      for (const packageSources of this.dbtRepository.packageToSourcesMap.entries()) {
+        for (const sourceTables of packageSources[1].entries()) {
+          const label = `(${packageSources[0]}) ${sourceTables[0]}`;
+          const insertText = this.getSourceInsertText(sourceTables[0], lastChar);
+          completionItems.push(this.getSourceCompletionItem(label, insertText, 'Source'));
+        }
+      }
+
+      return Promise.resolve(completionItems);
```
Comment (pgrivachev) [inline server/src/completion/SourceCompletionProvider.ts:25]: `return completionItems;`

File: server/src/completion/SourceCompletionProvider.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    const tableMatch = SourceCompletionProvider.TABLE_PATTERN.exec(jinjaBeforePositionText);
+    if (tableMatch) {
+      const sourceName = tableMatch[1].slice(1, -1);
+      return Promise.resolve(
```
Comment (pgrivachev) [inline server/src/completion/SourceCompletionProvider.ts:31]: The same.

File: server/src/definition/JinjaDefinitionProvider.ts [hunk 10]
```diff
@@ -15,32 +15,36 @@ export class JinjaDefinitionProvider {
 
   constructor(private dbtRepository: DbtRepository) {}
 
-  onJinjaDefinition(document: TextDocument, packageName: string | undefined, jinja: ParseNode, position: Position): DefinitionLink[] | undefined {
-    if (packageName && this.isExpression(jinja.value)) {
-      const refDefinitions = this.refDefinitionFinder.searchRefDefinitions(document, position, jinja, packageName, this.dbtRepository.models);
-      if (refDefinitions) {
-        return refDefinitions;
-      }
+  onJinjaDefinition(
```
Comment (pgrivachev) [inline server/src/definition/JinjaDefinitionProvider.ts:18]: This method can be refactored and looks similar to `DbtCompletionProvider` as the result. For definitionFinders we can have common interface.

File: README.md [hunk 11]
```diff
@@ -31,17 +31,18 @@ Also, it provides additional features like [Highlighting errors](#highlighting-e
...
 | [Completion for SQL](#completion-for-sql) | [BigQuery](https://docs.getdbt.com/reference/warehouse-profiles/bigquery-profile) |
+| [Completion for dbt](#completion-for-dbt) | All |
 | [Completion for dbt models](#completion-for-dbt-models) | [BigQuery](https://docs.getdbt.com/reference/warehouse-profiles/bigquery-profile) |
 | [dbt compile preview](#dbt-compile-preview) | All |
 | [Information on hover](#information-on-hover) | [BigQuery](https://docs.getdbt.com/reference/warehouse-profiles/bigquery-profile) |
 
 ### Go to definition
 
-#### Macros
+##### Macros
```
Comment (pgrivachev) [inline README.md:41]: Why did you increase count of #?

**Review by pgrivachev — COMMENTED** [review]
It looks good, we just need some refactor and unit tests.

File: server/src/test/completion/SourceCompletionProvider.spec.ts [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+  });
+
+  it(`Should provide sources by pressing '`, async () => {
+    await shouldProvideCompletions(sourceCompletionProvider, `select * from {{ source('`, [
+      { label: '(project_package) source_1', insertText: `source_1` },
+      { label: '(installed_package) package_source_1', insertText: `package_source_1` },
+    ]);
+  });
+
+  it('Should provide source tables', async () => {
```
Comment (pgrivachev) [inline server/src/test/completion/SourceCompletionProvider.spec.ts:73]: Also we can add tests for `undefined`.

File: server/src/test/helper.ts [hunk 13]
```diff
@@ -58,3 +60,21 @@ export function sleep(ms: number): Promise<unknown> {
     setTimeout(resolve, ms);
   });
 }
+
+export async function shouldProvideCompletions(
+  completionProvider: DbtNodeCompletionProvider,
+  text: string,
+  expectedCompletions: { label: string; insertText: string }[],
```
Comment (pgrivachev) [inline server/src/test/helper.ts:67]: Can we make it the same type as actual completions? `CompletionItem[] | undefined`

File: server/src/test/helper.ts [hunk 14]
```diff
@@ -58,3 +60,21 @@ export function sleep(ms: number): Promise<unknown> {
...
+): Promise<void> {
+  const completions = await completionProvider.provideCompletions(text);
+  assertCompletions(completions, expectedCompletions);
+}
+
+function assertCompletions(actualCompletions: CompletionItem[] | undefined, expectedCompletions: { label: string; insertText: string }[]): void {
```
Comment (pgrivachev) [inline server/src/test/helper.ts:73]: `expectedCompletions: CompletionItem[] | undefined`

File: server/src/test/helper.ts [hunk 15]
```diff
@@ -58,3 +60,21 @@ export function sleep(ms: number): Promise<unknown> {
...
+  assertThat(actualCompletions, defined());
+  actualCompletions?.forEach((actualItem, i) => {
```
Comment (pgrivachev) [inline server/src/test/helper.ts:75]: Can we use `contains` or `containsInAnyOrder`?

**Review by pgrivachev — COMMENTED** [review]
Thank you for adding these tests!

File: README.md (hunk 11 above)
Comment (fivetran-dmitryboykov) [reply README.md:41]: Fixed.

File: server/src/CompletionProvider.ts [hunk 16]
```diff
@@ -242,20 +241,6 @@ export class CompletionProvider {
 
   static readonly DBT_KEYWORDS = ['ref'];
```
[the remark this answers is not in the document]
Comment (fivetran-dmitryboykov) [reply server/src/CompletionProvider.ts:242]: Fixed.

File: server/src/DbtRepository.ts [hunk 17]
```diff
@@ -22,7 +22,75 @@ export class DbtRepository {
...
+
+  get packageToModelsMap(): Map<string, Set<ManifestModel>> {
+    return this.packageToModels;
+  }
+
+  get packageToMacrosMap(): Map<string, Set<ManifestMacro>> {
+    return this.packageToMacros;
+  }
+
+  get packageToSourcesMap(): Map<string, Map<string, Set<ManifestSource>>> {
```
[the remark this answers is not in the document]
Comment (fivetran-dmitryboykov) [reply server/src/DbtRepository.ts:42]: Seems better not to delete getters - I don't want to allow to update collections outside.

File: server/src/LspServer.ts [hunk 18]
```diff
@@ -79,7 +81,8 @@ export class LspServer {
   constructor(private connection: _Connection) {
     this.progressReporter = new ProgressReporter(this.connection);
     this.fileChangeListener = new FileChangeListener(this.yamlParser, this.manifestParser, this.dbtRepository);
-    this.completionProvider = new CompletionProvider(this.dbtRepository);
+    this.completionProvider = new CompletionProvider();
+    this.dbtCompletionProvider = new DbtCompletionProvider(this.dbtRepository);
```
[the remark this answers is not in the document]
Comment (fivetran-dmitryboykov) [reply server/src/LspServer.ts]: Seems that we don't need to hold multiple instances of them.

File: e2e/src/index.ts [hunk 19]
```diff
@@ -24,7 +24,7 @@ export async function run(): Promise<void> {
   return new Promise((resolve, reject) => {
     const startTime = performance.now();
 
-    glob('**.spec.js', { cwd: testsRoot }, (e, files) => {
+    glob('**/*.spec.js', { cwd: testsRoot }, (e, files) => {
```
[the remark this answers is not in the document]
Comment (fivetran-dmitryboykov) [reply e2e/src/index.ts:27]: Yes.

File: e2e/src/completion/completion_macros.spec.ts [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+  test('Should suggest macros', async () => {
+    // arrange
+    const docUri = getCustomDocUri(PROJECT_FILE_NAME);
+    await activateAndWait(docUri);
+
+    // act
+    const actualCompletionList = await triggerCompletion(docUri, new Position(0, 15), 'e');
+
+    // assert
+    const expectedCompletions = getMacrosCompletionList();
```
[the remark this answers is not in the document]
Comment (fivetran-dmitryboykov) [reply e2e/src/completion/completion_macros.spec.ts:23]: Seems not. VSCode returns a bit different completion item kinds for macros.

File: server/src/test/completion/ModelCompletionProvider.spec.ts [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+      { label: installedPackageModel1, insertText: `installed_package', 'installed_package_model_1` },
+      { label: installedPackageModel2, insertText: `installed_package', 'installed_package_model_2` },
+    ]);
+  });
+
+  it(`Should provide completions from all packages after pressing "`, async () => {
+    await shouldProvideCompletions(modelCompletionProvider, `select * from {{ ref("`, [
+      { label: labelModel1, insertText: 'model_1' },
+      { label: labelModel2, insertText: 'model_2' },
+      { label: installedPackageModel1, insertText: `installed_package", "installed_package_model_1` },
```
[the remark this answers is not in the document]
Comment (fivetran-dmitryboykov) [reply server/src/test/completion/ModelCompletionProvider.spec.ts:93]: Yes, seems that we can use textEdit field in CompletionItem.

File: server/src/completion/ModelCompletionProvider.ts [hunk 22]
```diff
@@ -51,9 +54,22 @@ export class ModelCompletionProvider implements DbtNodeCompletionProvider {
...
   private getModelInsertText(packageName: string, name: string, lastChar: string): string {
+    const lastCharIsQuote = isQuote(lastChar);
+
     if (this.dbtRepository.projectName === packageName) {
-      return lastChar === "'" ? `${name}` : `'${name}'`;
+      return new StringBuilder().appendIf(!lastCharIsQuote, "'").append(name).appendIf(!lastCharIsQuote, "'").toString();
     }
-    return lastChar === "'" ? `${packageName}', '${name}` : `'${packageName}', '${name}'`;
+
+    return new StringBuilder()
```
[the remark this answers is not in the document]
Comment (fivetran-dmitryboykov) [reply server/src/completion/ModelCompletionProvider.ts:63]: TBD.

File: server/src/completion/SourceCompletionProvider.ts [hunk 23]
```diff
@@ -51,6 +53,7 @@ export class SourceCompletionProvider implements DbtNodeCompletionProvider {
   }
 
   private getSourceInsertText(sourceName: string, lastChar: string): string {
-    return lastChar === "'" ? sourceName : `'${sourceName}'`;
+    const lastCharIsQuote = isQuote(lastChar);
+    return new StringBuilder().appendIf(!lastCharIsQuote, "'").append(sourceName).appendIf(!lastCharIsQuote, "'").toString();
```
[the remark this answers is not in the document]
Comment (fivetran-dmitryboykov) [reply server/src/completion/SourceCompletionProvider.ts:57]: TBD.

File: server/src/test/helper.ts (hunk 15 above)
Comment (fivetran-dmitryboykov) [reply server/src/test/helper.ts:75]: Seems complicated.

File: server/src/test/completion/ModelCompletionProvider.spec.ts [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+  it(`Should provide completions from all packages after pressing '`, async () => {
+    await shouldProvideCompletions(modelCompletionProvider, `select * from {{ ref('`, [
+      { label: labelModel1, insertText: 'model_1' },
+      { label: labelModel2, insertText: 'model_2' },
+      { label: installedPackageModel1, insertText: `installed_package', 'installed_package_model_1` },
+      { label: installedPackageModel2, insertText: `installed_package', 'installed_package_model_2` },
+    ]);
+  });
+
+  it(`Should provide completions from all packages after pressing "`, async () => {
```
[the remark this answers is not in the document]
Comment (fivetran-dmitryboykov) [reply server/src/test/completion/ModelCompletionProvider.spec.ts:89]: Seems not - usually we return more choices that expected. And VSCode filters them out.

File: server/src/test/completion/SourceCompletionProvider.spec.ts [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+      },
+    ];
+    dbtRepository.groupManifestNodes();
+
+    sourceCompletionProvider = new SourceCompletionProvider(dbtRepository);
+  });
+
+  it('Should provide sources by pressing (', async () => {
+    await shouldProvideCompletions(sourceCompletionProvider, `select * from {{ source(`, [
+      { label: '(project_package) source_1', insertText: `'source_1'` },
```
[the remark this answers is not in the document]
Comment (fivetran-dmitryboykov) [reply server/src/test/completion/SourceCompletionProvider.spec.ts:61]: Seems yes. Will need to research an ability to add custom configurations.

**Review by pgrivachev — APPROVED** [review]
Great!

File: server/src/completion/ModelCompletionProvider.ts [hunk 26]
```diff
@@ -54,22 +54,20 @@ export class ModelCompletionProvider implements DbtNodeCompletionProvider {
...
+    const isQuoteProvided = isQuote(lastChar);
+    const quoteSymbol = isQuoteProvided ? lastChar : `'`;
 
     if (this.dbtRepository.projectName === packageName) {
-      return new StringBuilder().appendIf(!lastCharIsQuote, "'").append(name).appendIf(!lastCharIsQuote, "'").toString();
+      return new StringBuilder().append(name).wrapIf(!isQuoteProvided, quoteSymbol).toString();
     }
 
     return new StringBuilder()
-      .appendIf(!lastCharIsQuote, "'")
```
[the remark this answers is not in the document]
Comment (fivetran-dmitryboykov) [reply server/src/completion/ModelCompletionProvider.ts]: Now we have `.wrapIf(` at the end.

File: e2e/src/completion/completion_models.spec.ts [hunk 27]
```diff
@@ -7,6 +7,13 @@ import { activateAndWait, getCustomDocUri, getTextInQuotesIfNeeded, setTestConte
 suite('Should suggest model completions', () => {
   const PROJECT_FILE_NAME = 'completion-jinja/models/completion_jinja.sql';
 
+  const MODELS_COMPLETIONS: [string, string][] = [
```
[the remark this answers is not in the document]
Comment (fivetran-dmitryboykov) [reply e2e/src/completion/completion_models.spec.ts:10]: Linter error after removing.

**Pull request merged**